### PR TITLE
Secure websocket

### PR DESCRIPTION
**Pull request by rshiv**
This PR adds the ability to integrate secure websockets to wakunode and chat2 . 
The tls key and certificate path is configurable using respective configs.

Increment towards  https://github.com/status-im/nim-waku/issues/434

### CODE REVIEW

File: examples/v2/config_chat2.nim [hunk 1]
```diff
@@ -225,6 +225,11 @@ type
       desc: "WebSocket listening port."
       defaultValue: 8000
       name: "websocket-port" }: Port
+    
+    websocketSecureSupport* {.
+      desc: "WebSocket Secure Support."
+      defaultValue: false
+      name: "websocket-Secure-Support" }: bool
```
Comment (jm-clius) [inline examples/v2/config_chat2.nim:232]:
```suggestion
      name: "websocket-secure-support" }: bool
```

Comment (jm-clius) [reply examples/v2/config_chat2.nim:232]: We can probably change the websocket enable/disable config items to simply `websocket` and `websocket-secure`?

**Review by jm-clius — COMMENTED** [review]
Thanks. At first glance this looks good! :) Will do a proper review after the weekend. I notice that you have a versioned/commited default key and certificate. Wouldn't we want people to be forced to provide their own - i.e. no default values for the config items? Not sure what behaviour we want. It makes sense to have example certificates in source control for our tests.

**Comment by rshiv** [comment]
> Thanks. At first glance this looks good! :) Will do a proper review after the weekend. I notice that you have a versioned/commited default key and certificate. Wouldn't we want people to be forced to provide their own - i.e. no default values for the config items? Not sure what behaviour we want. It makes sense to have example certificates in source control for our tests.

Exactly, I will move the demo path to test folder.

File: tests/v2/test_wakunode.nim [hunk 2]
```diff
@@ -27,9 +27,13 @@ when defined(rln):
   import ../../waku/v2/protocol/waku_rln_relay/[waku_rln_relay_utils, waku_rln_relay_types]
 
 const RLNRELAY_PUBSUB_TOPIC = "waku/2/rlnrelay/proto"
+template sourceDir: string = currentSourcePath.parentDir()
+const KEY_PATH = sourceDir / "../../waku/v2/node/key.txt"
```
Comment (oskarth) [inline tests/v2/test_wakunode.nim:31]: Probably don't want to store the key/cert (a) in node code folder (b) as txt.

Are these just for test?

File: tests/v2/test_wakunode.nim [hunk 3]
```diff
@@ -1164,4 +1168,107 @@ asyncTest "Messages relaying fails with non-overlapping transports (TCP or Webso
...
+    await node2.stop()
+
+asyncTest "Messages fails with wrong key path":
+    let
+      nodeKey1 = crypto.PrivateKey.random(Secp256k1, rng[])[]
+  
+    expect IOError:
+      # gibberish
+      discard WakuNode.new(nodeKey1, ValidIpAddress.init("0.0.0.0"),
+        bindPort = Port(60000), wsBindPort = Port(8000), wssEnabled = true, secureKey = "../../waku/v2/node/key_dummy.txt")
```
Comment (oskarth) [inline tests/v2/test_wakunode.nim:1224]: How is this different from key above?

File: waku/v2/node/config.nim [hunk 4]
```diff
@@ -247,6 +247,21 @@ type
...
       name: "websocket-port" }: Port
+    
+    websocketSecureSupport* {.
+      desc: "Enable secure websocket:  true|false",
+      defaultValue: false
+      name: "websocket-secure-support"}: bool
+    
+    websocketSecureKey* {.
+      desc: "Secure websocket key path:   '/path/to/key.txt' ",
+      defaultValue: "key.txt"
```
Comment (oskarth) [inline waku/v2/node/config.nim:258]: Having a default value here seems insecure. If we don't support auto generating it, I'd prefer if we we treated websocketSecureSupport=true with no key/cert as an error.

File: waku/v2/node/key.txt [hunk 5]
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
Comment (oskarth) [inline waku/v2/node/key.txt:1]: Should be in different folder - static perhaps? Also file extension seems off.

Assuming this is for tests only, we should make it clear that this is a dummy key for tests only.

File: waku/v2/utils/wakuswitch.nim [hunk 6]
```diff
@@ -14,6 +15,40 @@ import
 proc withWsTransport*(b: SwitchBuilder): SwitchBuilder =
   b.withTransport(proc(upgr: Upgrade): Transport = WsTransport.new(upgr))
 
+proc getSecureKey(path : string): TLSPrivateKey
+  {.raises: [Defect,TLSStreamProtocolError, IOError].} =
+  var stringkey : string = readFile(path)
+  ##echo stringkey
```
Comment (oskarth) [inline waku/v2/utils/wakuswitch.nim:21]: leftover

Comment (oskarth) [reply waku/v2/utils/wakuswitch.nim:21]: might be good to do a debug log here though

File: waku/v2/utils/wakuswitch.nim [hunk 7]
```diff
@@ -49,6 +87,9 @@ proc newWakuSwitch*(
     if wsEnabled == true:
       b = b.withAddresses(@[wsAddress, address])
       b = b.withWsTransport()
+    if wssEnabled == true:
```
Comment (oskarth) [inline waku/v2/utils/wakuswitch.nim:90]: Should we warn if someone enables wss and ws at same time? Not sure what is most common here. Similar to "https only"

**Review by oskarth — COMMENTED** [review]
Nice! Coming along nicely.

I'd also add a brief doc to documentation folder showing how to generate a cert w manually with letsencrypt and supply as command line argument.

This can also be raised as a Github Issue and done as a separate PR.

File: tests/v2/test_wakunode.nim (hunk 2 above)
Comment (rshiv) [reply tests/v2/test_wakunode.nim:31]: yes

File: tests/v2/test_wakunode.nim (hunk 3 above)
Comment (rshiv) [reply tests/v2/test_wakunode.nim:1224]: its failure case so , just bad path case.

File: waku/v2/utils/peers.nim [hunk 8]
```diff
@@ -82,12 +82,14 @@ proc parseRemotePeerInfo*(address: string): RemotePeerInfo {.raises: [Defect, Va
       p2pPart = addrPart.tryGet()
     of "ws":
       wsPart = addrPart.tryGet()
+    of "wss":
+      wsPart = addrPart.tryGet()
```
Comment (jm-clius) [inline waku/v2/utils/peers.nim:86]:
```suggestion
      wssPart = addrPart.tryGet()
```

File: waku/v2/utils/wakuswitch.nim [hunk 9]
```diff
@@ -14,6 +15,40 @@ import
 proc withWsTransport*(b: SwitchBuilder): SwitchBuilder =
   b.withTransport(proc(upgr: Upgrade): Transport = WsTransport.new(upgr))
 
+proc getSecureKey(path : string): TLSPrivateKey
+  {.raises: [Defect,TLSStreamProtocolError, IOError].} =
+  var stringkey : string = readFile(path)
+  ##echo stringkey
+  try :
+    let key: TLSPrivateKey = TLSPrivateKey.init(stringkey)
```
Comment (jm-clius) [inline waku/v2/utils/wakuswitch.nim:23]:
```suggestion
    let key = TLSPrivateKey.init(stringkey)
```
<-- Minor suggestion.
There's no need to specify the `type` when it's obvious from initialisation and we generally don't in `nim-waku.

File: waku/v2/utils/wakuswitch.nim [hunk 10]
```diff
@@ -14,6 +15,40 @@ import
...
+    raise newException(TLSStreamProtocolError,"Secure key init failed")
+    
+
+
+proc getSecureCert(path : string): TLSCertificate
+  {.raises: [Defect,TLSStreamProtocolError, IOError].} =
+  var
+    stringCert : string = readFile(path)
+  try :
+    let cert : TLSCertificate = TLSCertificate.init(stringCert)
```
Comment (jm-clius) [inline waku/v2/utils/wakuswitch.nim:35]:
```suggestion
    let cert  = TLSCertificate.init(stringCert)
```
See comment about type initialisation. Where obvious, no need to specify type.

File: waku/v2/node/wakunode2.nim [hunk 11]
```diff
@@ -130,25 +130,32 @@ template tcpEndPoint(address, port): auto =
...
   ## Creates a Waku Node.
   ##
   ## Status: Implemented.
   ##
   let
     rng = crypto.newRng()
     hostAddress = tcpEndPoint(bindIp, bindPort)
-    wsHostAddress = tcpEndPoint(bindIp, wsbindPort) & addWsFlag
+    wsHostAddress = if wssEnabled : tcpEndPoint(bindIp, wsbindPort) & addWssFlag
+                    else : tcpEndPoint(bindIp, wsbindPort) & addWsFlag
```
Comment (jm-clius) [inline waku/v2/node/wakunode2.nim:156]:
```suggestion
    wsHostAddress = if wssEnabled: tcpEndPoint(bindIp, wsbindPort) & addWssFlag
                    else: tcpEndPoint(bindIp, wsbindPort) & addWsFlag
```

Minor, but consistent use of `:` without leading whitespace. Elsewhere too.

File: waku/v2/node/wakunode2.nim [hunk 12]
```diff
@@ -130,25 +130,32 @@ template tcpEndPoint(address, port): auto =
...
     announcedAddresses = if extIp.isNone() or extPort.isNone(): @[]
-                        elif wsEnabled == false: @[tcpEndPoint(extIp.get(), extPort.get())]
+                        elif wsEnabled == false and wssEnabled == false: @[tcpEndPoint(extIp.get(), extPort.get())]
                         else : @[tcpEndPoint(extIp.get(), extPort.get()),
                         tcpEndPoint(extIp.get(), wsBindPort) & addWsFlag]
```
Comment (jm-clius) [inline waku/v2/node/wakunode2.nim:160]: Indentation. Bring this line in line with `tcpEndPoint` above so it's clearer to read as the next item in the `seq`.

Comment (jm-clius) [inline waku/v2/node/wakunode2.nim:160]: Why not just use the `wsHostAddress` you created above? It seems like you're doing the same thing here, but this won't include `wssFlag` when enabled.

**Review by jm-clius — CHANGES_REQUESTED** [review]
Looks good! Some minor comments and a fix below.

File: waku/v2/node/wakunode2.nim (hunk 12 above)
Comment (rshiv) [reply waku/v2/node/wakunode2.nim]: We cannot use wsHostAddress as bindIp maybe different for extIP.

File: waku/v2/utils/wakuswitch.nim (hunk 7 above)
Comment (rshiv) [reply waku/v2/utils/wakuswitch.nim:93]: added debug log

File: waku/v2/node/wakunode2.nim (hunk 12 above)
Comment (rshiv) [reply waku/v2/node/wakunode2.nim]: though announced address was missing wss flag , added the same.

File: tests/v2/test_wakunode.nim [hunk 13]
```diff
@@ -1164,4 +1168,107 @@ asyncTest "Messages relaying fails with non-overlapping transports (TCP or Webso
...
+    check:
+      (await completionFut.withTimeout(5.seconds)) == true
+    await node1.stop()
+    await node2.stop()
+
+
+    
+    
+
+
```
Comment (staheri14) [inline tests/v2/test_wakunode.nim:1274]:
```suggestion
```

File: waku/v2/node/wakunode2.nim [hunk 14]
```diff
@@ -130,35 +130,46 @@ template tcpEndPoint(address, port): auto =
...
+                          tcpEndPoint(extIp.get(), wsBindPort) & addWsFlag]
     peerInfo = PeerInfo.new(nodekey)
     enrIp = if extIp.isSome(): extIp
             else: some(bindIp)
     enrTcpPort = if extPort.isSome(): extPort
                  else: some(bindPort)
     enr = createEnr(nodeKey, enrIp, enrTcpPort, none(Port))
   
-  if wsEnabled == true:
+  if wsEnabled == true or wssEnabled == true:
```
Comment (staheri14) [inline waku/v2/node/wakunode2.nim:172]: Can there be a case that `wsEnabled` is false and `wssEnabled` true?

File: waku/v2/node/config.nim [hunk 15]
```diff
@@ -247,6 +247,21 @@ type
...
+      desc: "Enable secure websocket:  true|false",
+      defaultValue: false
+      name: "websocket-secure-support"}: bool
+    
+    websocketSecureKey* {.
+      desc: "Secure websocket key path:   '/path/to/key.txt' ",
+      defaultValue: ""
+      name: "websocket-secure-key"}: string
+    
+    websocketSecureCert* {.
```
Comment (staheri14) [inline waku/v2/node/config.nim:261]: What do you think about changing it to `websocketSecureCertPath` to be more accurate? 
ditto `websocketSecureKey`

File: waku/v2/utils/wakuswitch.nim [hunk 16]
```diff
@@ -14,6 +15,40 @@ import
 proc withWsTransport*(b: SwitchBuilder): SwitchBuilder =
   b.withTransport(proc(upgr: Upgrade): Transport = WsTransport.new(upgr))
 
+proc getSecureKey(path : string): TLSPrivateKey
+  {.raises: [Defect,TLSStreamProtocolError, IOError].} =
+  trace "Key path is.", path=path
+  var stringkey : string = readFile(path)
+  try :
```
Comment (staheri14) [inline waku/v2/utils/wakuswitch.nim:22]:
```suggestion
  try:
```

File: waku/v2/utils/wakuswitch.nim [hunk 17]
```diff
@@ -14,6 +15,40 @@ import
 proc withWsTransport*(b: SwitchBuilder): SwitchBuilder =
   b.withTransport(proc(upgr: Upgrade): Transport = WsTransport.new(upgr))
 
+proc getSecureKey(path : string): TLSPrivateKey
+  {.raises: [Defect,TLSStreamProtocolError, IOError].} =
+  trace "Key path is.", path=path
+  var stringkey : string = readFile(path)
```
Comment (staheri14) [inline waku/v2/utils/wakuswitch.nim:21]:
```suggestion
  var stringkey: string = readFile(path)
```

File: waku/v2/utils/wakuswitch.nim [hunk 18]
```diff
@@ -14,6 +15,40 @@ import
...
+  except:
+    raise newException(TLSStreamProtocolError,"Secure key init failed")
+    
+
+
+proc getSecureCert(path : string): TLSCertificate
+  {.raises: [Defect,TLSStreamProtocolError, IOError].} =
+  trace "Certificate path is.", path=path
+  var stringCert : string = readFile(path)
+  try :
```
Comment (staheri14) [inline waku/v2/utils/wakuswitch.nim:34]:
```suggestion
  var stringCert: string = readFile(path)
  try:
```

File: tests/v2/test_wakunode.nim [hunk 19]
```diff
@@ -27,9 +27,13 @@ when defined(rln):
   import ../../waku/v2/protocol/waku_rln_relay/[waku_rln_relay_utils, waku_rln_relay_types]
 
 const RLNRELAY_PUBSUB_TOPIC = "waku/2/rlnrelay/proto"
+template sourceDir: string = currentSourcePath.parentDir()
+const KEY_PATH = sourceDir / "resources/key.pem"
```
Comment (oskarth) [inline tests/v2/test_wakunode.nim:31]: Sorry to be nitpicky but can we preface this with `test` like `test_key`? We really don't want people to accidentally end up using it in production. Ditto cert.

**Review by oskarth — APPROVED** [review]
Nice! Think good enough to be merged and tested .Minor comment. Also pending Hanno/Sanaz review

File: waku/v2/node/wakunode2.nim (hunk 14 above)
Comment (rshiv) [reply waku/v2/node/wakunode2.nim:172]: If the user choses to do so , we will raise a warning however the logic for wakunode won't change as we will take wss as its the secure version

File: tests/v2/test_wakunode.nim (hunk 19 above)
Comment (rshiv) [reply tests/v2/test_wakunode.nim]: changed in last commit

**Review by staheri14 — APPROVED** [review]
LGTM!

**Review by jm-clius — APPROVED** [review]
Thanks!

**Comment by jm-clius** [comment]
One more comment - I notice that `jenkins/prs/windows` is failing. Might be worth investigating/ensuring that it's expected before merging. I know we have known failures in `jenkins/prs/linux` and `Test / windows`.

**Comment by rshiv** [comment]
> One more comment - I notice that `jenkins/prs/windows` is failing. Might be worth investigating/ensuring that it's expected before merging. I know we have known failures in `jenkins/prs/linux` and `Test / windows`.

I checked the same, seems v1 test also failing in the jenkins/windows build, seems jenkins issue.

**Pull request merged**